File: typescript/src/tools/filesystem.ts
import fs from "fs";
import path from "path";

export function readFile(args: { path: string }): string {
  if (!fs.existsSync(args.path)) return `ERROR: ${args.path} does not exist`;
  if (!fs.statSync(args.path).isFile()) return `ERROR: ${args.path} is not a file`;
  try {
    return fs.readFileSync(args.path, "utf-8");
  } catch (e) {
    return `ERROR: ${String(e)}`;
  }
}

export function writeFile(args: { path: string; content: string }): string {
  try {
    fs.mkdirSync(path.dirname(args.path), { recursive: true });
    fs.writeFileSync(args.path, args.content, "utf-8");
    return `OK: wrote ${args.content.length} bytes to ${args.path}`;
  } catch (e) {
    return `ERROR: ${String(e)}`;
  }
}

export function appendFile(args: { path: string; content: string }): string {
  try {
    fs.mkdirSync(path.dirname(args.path), { recursive: true });
    fs.appendFileSync(args.path, args.content, "utf-8");
    return `OK: appended ${args.content.length} bytes to ${args.path}`;
  } catch (e) {
    return `ERROR: ${String(e)}`;
  }
}

function walkDir(dir: string, base: string): string[] {
  const entries: string[] = [];
  try {
    for (const e of fs.readdirSync(dir)) {
      if (e.startsWith(".")) continue;
      const full = path.join(dir, e);
      const rel = path.relative(base, full);
      if (fs.statSync(full).isDirectory()) {
        entries.push(...walkDir(full, base));
      } else {
        entries.push(rel);
      }
    }
  } catch { /* skip unreadable dirs */ }
  return entries;
}

export function listDirectory(args: { path: string; recursive?: boolean }): string {
  if (!fs.existsSync(args.path)) return `ERROR: ${args.path} does not exist`;
  if (!fs.statSync(args.path).isDirectory()) return `ERROR: not a directory`;
  try {
    if (args.recursive) return walkDir(args.path, args.path).join("\n") || "(empty)";
    const entries = fs.readdirSync(args.path).map((e) => {
      return fs.statSync(path.join(args.path, e)).isDirectory() ? e + "/" : e;
    });
    return entries.join("\n") || "(empty)";
  } catch (e) {
    return `ERROR: ${String(e)}`;
  }
}

export function createDirectory(args: { path: string }): string {
  try {
    fs.mkdirSync(args.path, { recursive: true });
    return `OK: created ${args.path}`;
  } catch (e) {
    return `ERROR: ${String(e)}`;
  }
}

export function deleteFile(args: { path: string }): string {
  if (!fs.existsSync(args.path)) return `ERROR: ${args.path} does not exist`;
  try {
    if (fs.statSync(args.path).isDirectory()) {
      fs.rmSync(args.path, { recursive: true, force: true });
      return `OK: deleted directory ${args.path}`;
    }
    fs.unlinkSync(args.path);
    return `OK: deleted ${args.path}`;
  } catch (e) {
    return `ERROR: ${String(e)}`;
  }
}

export function moveFile(args: { source: string; dest: string }): string {
  if (!fs.existsSync(args.source)) return `ERROR: ${args.source} does not exist`;
  try {
    fs.mkdirSync(path.dirname(args.dest), { recursive: true });
    fs.renameSync(args.source, args.dest);
    return `OK: moved ${args.source} → ${args.dest}`;
  } catch (e) {
    return `ERROR: ${String(e)}`;
  }
}

export function searchFiles(args: { pattern: string; directory: string; file_glob?: string }): string {
  if (!fs.existsSync(args.directory)) return `ERROR: ${args.directory} does not exist`;
  const results: string[] = [];

  let regex: RegExp;
  try {
    regex = new RegExp(args.pattern);
  } catch {
    regex = new RegExp(args.pattern.replace(/[.*+?^${}()|[\]\\]/g, "\\$&"));
  }

  const glob = args.file_glob ?? "*";
  const globRegex = new RegExp("^" + glob.replace(/\./g, "\\.").replace(/\*/g, ".*") + "$");

  function walk(dir: string) {
    let entries: string[];
    try { entries = fs.readdirSync(dir); } catch { return; }
    for (const e of entries) {
      if (e.startsWith(".")) continue;
      const full = path.join(dir, e);
      let stat;
      try { stat = fs.statSync(full); } catch { continue; }
      if (stat.isDirectory()) { walk(full); continue; }
      if (!globRegex.test(e)) continue;
      try {
        const lines = fs.readFileSync(full, "utf-8").split("\n");
        for (let i = 0; i < lines.length; i++) {
          if (regex.test(lines[i])) {
            results.push(`${path.relative(args.directory, full)}:${i + 1}: ${lines[i].trim()}`);
          }
        }
      } catch { /* skip unreadable files */ }
    }
  }
  walk(args.directory);
  if (!results.length) return "No matches found";
  return results.slice(0, 200).join("\n");
}

File: python/src/redtonomous/tools/filesystem.py
import fnmatch
import os
import re
import shutil
from pathlib import Path


def read_file(path: str) -> str:
    p = Path(path)
    if not p.exists():
        return f"ERROR: {path} does not exist"
    if not p.is_file():
        return f"ERROR: {path} is not a file"
    try:
        return p.read_text(errors="replace")
    except Exception as e:
        return f"ERROR: {e}"


def write_file(path: str, content: str) -> str:
    try:
        p = Path(path)
        p.parent.mkdir(parents=True, exist_ok=True)
        p.write_text(content)
        return f"OK: wrote {len(content)} bytes to {path}"
    except Exception as e:
        return f"ERROR: {e}"


def append_file(path: str, content: str) -> str:
    try:
        p = Path(path)
        p.parent.mkdir(parents=True, exist_ok=True)
        with open(p, "a") as f:
            f.write(content)
        return f"OK: appended {len(content)} bytes to {path}"
    except Exception as e:
        return f"ERROR: {e}"


def list_directory(path: str, recursive: bool = False) -> str:
    p = Path(path)
    if not p.exists():
        return f"ERROR: {path} does not exist"
    if not p.is_dir():
        return f"ERROR: {path} is not a directory"
    if recursive:
        entries = []
        for root, dirs, files in os.walk(p):
            # skip hidden dirs
            dirs[:] = [d for d in dirs if not d.startswith(".")]
            rel = Path(root).relative_to(p)
            for f in files:
                entries.append(str(rel / f))
        return "\n".join(sorted(entries)) or "(empty)"
    entries = sorted(str(e.name) + ("/" if e.is_dir() else "") for e in p.iterdir())
    return "\n".join(entries) or "(empty)"


def create_directory(path: str) -> str:
    Path(path).mkdir(parents=True, exist_ok=True)
    return f"OK: created {path}"


def delete_file(path: str) -> str:
    p = Path(path)
    if not p.exists():
        return f"ERROR: {path} does not exist"
    if p.is_dir():
        shutil.rmtree(p)
        return f"OK: deleted directory {path}"
    p.unlink()
    return f"OK: deleted {path}"


def move_file(source: str, dest: str) -> str:
    s, d = Path(source), Path(dest)
    if not s.exists():
        return f"ERROR: {source} does not exist"
    d.parent.mkdir(parents=True, exist_ok=True)
    shutil.move(str(s), str(d))
    return f"OK: moved {source} → {dest}"


def search_files(pattern: str, directory: str, file_glob: str = "*") -> str:
    results = []
    root = Path(directory)
    if not root.exists():
        return f"ERROR: {directory} does not exist"
    try:
        regex = re.compile(pattern)
    except re.error as e:
        return f"ERROR: invalid regex {pattern!r}: {e}"

    for dirpath, dirs, files in os.walk(root):
        dirs[:] = [d for d in dirs if not d.startswith(".")]
        for fname in files:
            if not fnmatch.fnmatch(fname, file_glob):
                continue
            fpath = Path(dirpath) / fname
            try:
                for lineno, line in enumerate(fpath.read_text(errors="replace").splitlines(), 1):
                    if regex.search(line):
                        rel = fpath.relative_to(root)
                        results.append(f"{rel}:{lineno}: {line.strip()}")
            except Exception:
                continue
    if not results:
        return "No matches found"
    return "\n".join(results[:200])
